amazon/ion/ion_py_objects.py

```diff
@@ -0,0 +1,586 @@
+
+    @classmethod
+    def from_event(cls, ion_event):
+        if ion_event.value is not None:
+            args, kwargs = cls._to_constructor_args(ion_event.value)
+        else:
+            args, kwargs = (None, [], ()), {}
+        value = cls(*args, **kwargs)
+        value.ion_type = ion_event.ion_type
+        value.ion_annotations = ion_event.annotations
+        return value
+
+    @classmethod
+    def from_value(cls, ion_type, value, annotations=()):
+        if value is None:
+            value = IonPyNull_new()
+        else:
+            args, kwargs = cls._to_constructor_args(value)
+            value = cls(*args, **kwargs)
+        value.ion_type = ion_type
+        value.ion_annotations = annotations
+        return value
+
+    def to_event(self, event_type, field_name=None, in_struct=False, depth=None):
+        value = self
+        if isinstance(self, IonPyNull_new) or self.ion_type.is_container:
+            value = None
+
+        if in_struct:
+            if not isinstance(field_name, SymbolToken):
+                field_name = SymbolToken(field_name, 0 if field_name is None else None)
+        else:
+            field_name = None
+
+        return IonEvent(event_type, ion_type=self.ion_type, value=value, field_name=field_name,
+                        annotations=self.ion_annotations, depth=depth)
+
+
+class IonPyDict_new(MutableMapping):
+    __name__ = 'IonPyDict_new'
+    __qualname__ = 'IonPyDict_new'
+    ion_type = IonType.STRUCT
+
+    def __init__(self, ion_type=IonType.STRUCT, value=None, annotations=()):
+        super().__init__()
+        self.ion_annotations = annotations
+        self.__store = {}
+        if value is not None:
+            for key, value in iter(value.items()):
+                self.__store[key] = [value]
+
+    def __getitem__(self, key):
+        return self.__store[key][len(self.__store[key]) - 1]  # Return only one in order not to break clients
+
+    def __delitem__(self, key):
+        del self.__store[key]
+
+    def __setitem__(self, key, value):
+        self.__store[key] = [value]
+
+    def __len__(self):
+        return sum([len(values) for values in iter(self.__store.values())])
+
+    def __iter__(self):
+        for key in iter(self.__store.keys()):
+            yield key
+
+    def __str__(self):
+        return repr(self)
+
+    def __repr__(self):
+        return '{%s}' % ', '.join(['%r: %r' % (k, v) for k, v in self.items()])
+
+    def add_item(self, key, value):
+        if key in self.__store:
+            self.__store[key].append(value)
+        else:
+            self.__setitem__(key, value)
+
+    def get_all_values(self, key):
+        return self.__store[key]
+
+    def iteritems(self):
+        for key in self.__store:
+            for value in self.__store[key]:
+                yield key, value
+
+    def items(self):
+        output = []
+        for k, v in self.iteritems():
+            output.append((k, v))
+        return output
```

Comment:
why does this exist?

Reply:
Would you refer to `items()` vs.`iteritems()`?

Python 3 has deprecated `iteritems`. I removed `iteritems` in the coming commit too.

Reply:
The python 3 `dict.items()` returns a set-like object providing a view on the dictionary's items.

Our `IonPydict.item()` leverage the python generator to iterate the IonPyDict's items.

This Python 2/3 and dict/IonPyDict behavior difference need to be fixed or documented. I'll create a Github issue and see if we still need this method and why.

Reply:
Created GH issue - https://github.com/amazon-ion/ion-python/issues/288